### ISSUE TITLE
Unsatisfiable Indentation module scenario inside Lambda (Java 8)

### ISSUE THREAD

**Issue by dmitry-cherkas** [checkstyle/checkstyle#577]
It seems, there is an issue with indentation tracking when lambda is used to declare inner class, like this:

```java
public class ClassWithInnerLambda {

    Runnable someLambda = () -> {
        String varWithAppropriateIndention = null;
    };
}
```
then checkstyle incorrectly says
> [WARNING] ClassWithInnerLambda.java[4] (indentation) Indentation: 'String' have incorrect indentation level 8, expected level should be 12.

After indenting 'String' by 4 spaces, checkstyle correctly says
> [WARNING] ClassWithInnerLambda.java[4] (indentation) Indentation: 'block' child have incorrect indentation level 12, expected level should be 8.
> [WARNING] ClassWithInnerLambda.java[4] (indentation) Indentation: 'member def type' have incorrect indentation level 12, expected level should be 8.

Here is my checkstyle configuration:
```xml
<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <property name="severity" value="warning"/>

    <module name="TreeWalker">
        <module name="Indentation">
            <property name="basicOffset" value="4"/>
            <property name="braceAdjustment" value="0"/>
            <property name="caseIndent" value="4"/>
            <property name="throwsIndent" value="8"/>
            <property name="lineWrappingIndentation" value="8"/>
            <property name="arrayInitIndent" value="4"/>
        </module>
    </module>
</module>
```

**Comment by dmitry-cherkas** [comment]
I understand, you guys are busy out there, but may be you could give me a clue so that I could fix this issue myself? I have cloned the repo, created a test and I see, it fails on the following trace:
com/puppycrawl/tools/checkstyle/checks/indentation/MemberDefHandler.java:46
com/puppycrawl/tools/checkstyle/checks/indentation/MemberDefHandler.java:68
com/puppycrawl/tools/checkstyle/checks/indentation/LineWrappingHandler.java:154

ending up in com.puppycrawl.tools.checkstyle.checks.indentation.LineWrappingHandler#logWarningMessage

But I have no idea of how to fix it, or what is the root cause. Apparently, checkStyle fails to recognize Lambda statement like this, mistakenly treating it as long declaration. Please advice me of any further steps I should dig in.

Thanks in advance.

**Comment by romani** [comment]
Sorry, but we still do not have any support for lambda, https://github.com/checkstyle/checkstyle/issues/281 , that is the most complecated Check in checkstyle, it hard to fix it and explain.
@maxvetrenko , can you speedup your activity for lambda support or explain how we could help you?

**Comment by dmitry-cherkas** [comment]
Ah, I see. I assumed Lambdas are supported, since #10 is closed and there are some mentions of "Lambda" in checkstyle sources.

**Comment by romani** [comment]
https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java#L3437

by #10, we can parse java8 sources , but all Checks need to be updated separately to properly process them.

**Comment by dmitry-cherkas** [comment]
Ok, so if you state, that Abstract Syntax Tree, checkstyle is building, is aware of Lambdas and is built correctly in my case, the task of adding indention support is as easy as providing proper implementation of com.puppycrawl.tools.checkstyle.checks.indentation.ExpressionHandler which is mapped to com.puppycrawl.tools.checkstyle.api.TokenTypes.LAMBDA. At least, at the first glance :)

Am I right?

**Comment by romani** [comment]
yes , we need smth similar to https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/NewHandler.java but for LAMBDA

**Comment by rnveach** [comment]
@romani I cannot reproduce the original issue in the newest checkstyle (no audit errors) and it looks similar to some examples we have in `InputLambda1`.

**Comment by romani** [comment]
@rnveach , please prove this by CLI output and lets close the issue.

**Comment by rnveach** [comment]
@romani 
output from: `java -jar checkstyle-6.15-all.jar -c config.xml TestClass.java` using the config and test class above.
````
Starting audit...
Audit done.
````

**Comment by romani** [comment]
What is TestClass.java ?
please do this as http://checkstyle.sourceforge.net/report_issue.html#How_to_report_a_bug

**Comment by rnveach** [comment]
It is what I saved the file under. I always use that when I report issues.

````
$ javac TestClass.java

$ cat TestClass.java
public class TestClass {
    Runnable someLambda = () -> {
        String varWithAppropriateIndention = null;
    };
}

$ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <property name="severity" value="warning"/>

    <module name="TreeWalker">
        <module name="Indentation">
            <property name="basicOffset" value="4"/>
            <property name="braceAdjustment" value="0"/>
            <property name="caseIndent" value="4"/>
            <property name="throwsIndent" value="8"/>
            <property name="lineWrappingIndentation" value="8"/>
            <property name="arrayInitIndent" value="4"/>
        </module>
    </module>
</module>

$ java -jar checkstyle-6.15-all.jar -c config.xml TestClass.java
Starting audit...
Audit done.
````

**Comment by romani** [comment]
we can not reproduce the issue on latest release.
Issue is closed.
